File: pkg/processor/runtime/ruby/wrapper.rb
#
# Copyright 2017 The Nuclio Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'optparse'
require 'socket'
require 'json'
require 'base64'
require 'date'

class Logger
  def initialize(socket)
    @socket = socket
  end

  def debug(message, **with)
    log(:debug, message, **with)
  end

  def info(message, **with)
    log(:info, message, **with)
  end

  def warn(message, **with)
    log(:warning, message, **with)
  end

  def error(message, **with)
    log(:error, message, **with)
  end

  def log(level, message, **with)
    log_val = {
      level: level,
      message: message,
      with: with,
      datetime: Time.now.strftime('%Y-%m-%dT%H:%M:%S.%L%z')
    }
    @socket.puts "l#{log_val.to_json}"
  end
end

class Context
  attr_reader :logger

  def initialize(logger)
    @logger = logger
  end
end

class ByteBuffer
  def initialize(bytes)
    @bytes = bytes
  end

  def base64_encode
    Base64.encode64(@bytes)
  end
end

class KeywordStruct < Struct
  def initialize(**kwargs)
    super(*members.map { |k| kwargs[k] })
  end
end

Response = Struct.new(:body, :headers, :content_type, :status_code, :body_encoding) do
  def initialize(body, headers: {}, content_type: 'text/plain', status_code: 200, body_encoding: 'text')
    super(body, headers, content_type, status_code, body_encoding)
  end
end

Event = KeywordStruct.new(:body, :content_type, :headers, :fields, :id, :method, :path, :url, :timestamp, :trigger, :version)

Trigger = KeywordStruct.new(:class_name, :kind)

def response_from_output(handler_output)
  if handler_output.is_a?(Response)
    handler_output
  elsif handler_output.is_a?(Array) && handler_output.size == 2
    status_code = handler_output.first
    body, content_type, body_encoding = response_info_from_output(handler_output.last)
    Response.new(body, status_code: status_code, content_type: content_type, body_encoding: body_encoding)
  else
    body, content_type, body_encoding = response_info_from_output(handler_output)
    Response.new(body, status_code: 200, content_type: content_type, body_encoding: body_encoding)
  end
end

def response_info_from_output(handler_output)
  case handler_output
  when String
    [handler_output, 'text/plain', 'text']
  when ByteBuffer
    [handler_output.base64_encode, 'text/plain', 'base64']
  else
    [handler_output.to_json, 'application/json', 'text']
  end
end

def parse_event(input)
  json = JSON.parse(input)
  trigger = Trigger.new(class_name: json['trigger']['class'], kind: json['trigger']['kind'])
  Event.new(
    body: Base64.decode64(json['body']),
    content_type: json['content_type'],
    headers: json['headers'],
    fields: json['fields'],
    id: json['id'],
    method: json['method'],
    path: json['path'],
    url: json['url'],
    timestamp: DateTime.strptime(json['timestamp'].to_s, '%s'),
    trigger: trigger,
    version: json['version']
  )
end

if $PROGRAM_NAME == __FILE__
  options = {}
  OptionParser.new do |opt|
    opt.on('--handler HANDLER') { |o| options[:handler] = o }
    opt.on('--socket-path SOCKET_PATH') { |o| options[:socket_path] = o }
  end.parse!

  file, method_name = options[:handler].split(':')

  require_relative file

  socket = UNIXSocket.new(options[:socket_path])
  logger = Logger.new(socket)
  while input = socket.gets
    begin
      context = Context.new(logger)
      event = parse_event(input)
      res = send(method_name, context, event)
      encoded = response_from_output(res)
    rescue StandardError => e
      res = "#{e.backtrace.first}: #{e.message} (#{e.class})\n#{e.backtrace.drop(1).join("\n")}"
      encoded = Response.new(res, status_code: 500)
    end
    logger.debug('Response is', response: encoded.to_h)
    socket.puts "r#{encoded.to_h.to_json}"
  end
  socket.close
end

File: pkg/processor/runtime/nodejs/js/wrapper.js
/*
Copyright 2023 The Nuclio Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

const net = require('net')
const events = require('events')

const jsonCtype = 'application/json'
const initContextFunctionName = 'initContext'

const messageTypes = {
    LOG: 'l',
    RESPONSE: 'r',
    METRIC: 'm',
    START: 's',
}

const logLevels = {
    DEBUG: 'debug',
    INFO: 'info',
    WARNING: 'warning',
    ERROR: 'error',
}

let context = {
    userData: {},
    callback: async (handlerResponse) => {
        context._eventEmitter.emit('callback', handlerResponse)
    },
    Response: Response,
    logger: {
        error: logWithLevel(logLevels.ERROR),
        warn: logWithLevel(logLevels.WARNING),
        info: logWithLevel(logLevels.INFO),
        debug: logWithLevel(logLevels.DEBUG),
        errorWith: logWithLevel(logLevels.ERROR),
        warnWith: logWithLevel(logLevels.WARNING),
        infoWith: logWithLevel(logLevels.INFO),
        debugWith: logWithLevel(logLevels.DEBUG),
    },
    _socket: undefined,
    _eventEmitter: new events.EventEmitter(),
}

function Response(body = null,
                  headers = null,
                  contentType = 'text/plain',
                  statusCode = 200,
                  bodyEncoding = 'text') {
    this.body = body
    this.headers = headers
    this.content_type = contentType
    this.status_code = statusCode
    this.body_encoding = bodyEncoding

    if (!isString(this.body)) {
        this.body = JSON.stringify(this.body)
        this.content_type = jsonCtype
    }
}

function writeMessageToProcessor(messageType, messageContents) {
    context._socket.write(`${messageType}${messageContents}\n`)
}

function logWithLevel(level) {
    return (...args) => log(level, ...args)
}

function log(level, message, withData = {}) {
    const datetime = (new Date()).toISOString()
    const record = {
        datetime,
        level,
        message,
        with: withData,
    }
    writeMessageToProcessor(messageTypes.LOG, JSON.stringify(record))
}

function isString(obj) {
    return typeof (obj) === 'string' || (obj instanceof String)
}

// Status reply is a list of [status, content]
function isStatusReply(handlerOutput) {
    return Array.isArray(handlerOutput) &&
        handlerOutput.length === 2 &&
        typeof (handlerOutput[0]) === 'number'
}

function responseFromOutput(handlerOutput) {
    let response = {
        body: '',
        content_type: 'text/plain',
        headers: {},
        status_code: 200,
        body_encoding: 'text',
    }

    if (isString(handlerOutput) || (handlerOutput instanceof Buffer)) {
        response.body = handlerOutput
    } else if (isStatusReply(handlerOutput)) {
        response.status_code = handlerOutput[0]
        const body = handlerOutput[1]

        if (isString(body) || (body instanceof Buffer)) {
            response.body = body
        } else {
            response.body = JSON.stringify(body)
            response.content_type = jsonCtype
        }
    } else if (handlerOutput instanceof Response) {
        Object.assign(response, handlerOutput)
    } else {

        // other object
        response.body = JSON.stringify(handlerOutput)
        response.content_type = jsonCtype
    }

    if (Buffer.isBuffer(response.body)) {
        response.body = response.body.toString('base64')
        response.body_encoding = 'base64'
    }

    return response
}

function writeDuration(start, end) {
    const duration = {
        duration: Math.max(0.00000000001, (end.getTime() - start.getTime()) / 1000)
    }
    writeMessageToProcessor(messageTypes.METRIC, JSON.stringify(duration))
}

async function handleEvent(handlerFunction, incomingEvent) {
    let response = {}
    try {
        incomingEvent.body = new Buffer.from(incomingEvent['body'], 'base64')
        incomingEvent.timestamp = new Date(incomingEvent['timestamp'] * 1000)

        const start = new Date()

        // listening on response before executing, to avoid deadlock
        const responseWaiter = new Promise(resolve => context
            ._eventEmitter
            .once('callback', resolve))

        // call the handler
        handlerFunction(context, incomingEvent)

        // wait for callback
        const handlerResponse = await responseWaiter

        // write execution duration
        const end = new Date()
        writeDuration(start, end)
        response = responseFromOutput(handlerResponse)

    } catch (err) {
        console.log(`ERROR: ${err}`)
        let errorMessage = err.toString()

        if (err.stack !== undefined) {
            console.log(err.stack)
            errorMessage += `\n${err.stack}`
        }

        response = {
            body: `Error in handler: ${errorMessage}`,
            content_type: 'text/plain',
            headers: {},
            status_code: 500,
            body_encoding: 'text'
        }
    } finally {

        // write response
        writeMessageToProcessor(messageTypes.RESPONSE, JSON.stringify(response))
    }
}

function connectSocket(socketPath, handlerFunction) {
    const socket = new net.Socket()
    console.log(`socketPath = ${socketPath}`)
    if (socketPath.includes(':')) {

        // TCP - host:port
        const [host, portStr] = socketPath.split(':')
        const port = Number.parseInt(portStr)
        socket.connect(port, host)
    } else {

        // UNIX
        socket.connect(socketPath)
    }
    context._socket = socket
    socket.on('ready', () => {
        writeMessageToProcessor(messageTypes.START, '')
    })
    socket.on('data', async data => {
        let incomingEvent = JSON.parse(data)
        await handleEvent(handlerFunction, incomingEvent)
    })
}

function executeInitContext(functionModule) {
    const initContextFunction = functionModule[initContextFunctionName]
    if (typeof initContextFunction === 'function') {
        return initContextFunction(context)
    }
}

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms))
}

async function findFunction(functionModule, name) {

    // attempt to find the handler for a few seconds - connect if/when we do
    // if the handler wasn't found within the limit set here, give up
    let functionToFind = functionModule[name]
    let attempts = 0
    const delayMilliseconds = 250
    while (functionToFind === undefined) {
        if (attempts < 40) {
            console.warn(`function "${name}" not found yet in ${functionModule}`)
            attempts++
            await sleep(delayMilliseconds)
        } else {
            console.warn(`function "${name}" not found by deadline in ${functionModule}`)
            throw `Failed to find function "${name}" in "${functionModule}"`
        }
        functionToFind = functionModule[name]
    }
    return functionToFind
}

function run(socketPath, handlerPath, handlerName) {
    const functionModule = require(handlerPath)
    return findFunction(functionModule, handlerName)
        .then(async handlerFunction => {
            try {
                executeInitContext(functionModule)
            } catch (err) {
                console.error(`Failed to init context: ${err}`)
                throw err
            }
            return connectSocket(socketPath, handlerFunction)
        })
}

if (require.main === module) {

    // First two arguments are ['node', '/path/to/wrapper.js']
    const args = process.argv.slice(2)

    // ['/path/to/socket', '/path/to/handler.js', 'handler']
    if (args.length !== 3) {
        console.error('error: wrong number of arguments')
        process.exit(1)
    }

    const socketPath = args[0]
    const handlerPath = args[1]
    const handlerName = args[2]

    run(socketPath, handlerPath, handlerName)
        .catch((err) => {
            console.error('Error occurred during running. Error:', err)
            process.exit(1)
        })
}
